test/workers/BalanceRequestUpdateWorkerSpec.scala

```diff
@@ -0,0 +1,140 @@
+import play.api.test.FutureAwaits
+import repositories.BalanceRequestRepositoryImpl
+import services.BalanceRequestCacheService
+import uk.gov.hmrc.mongo.test.DefaultPlayMongoRepositorySupport
+import uk.gov.hmrc.play.bootstrap.config.ServicesConfig
+
+import java.security.SecureRandom
+import java.time.Clock
+import java.time.ZoneOffset
+
+class BalanceRequestUpdateWorkerSpec
+  extends AnyFlatSpec
+  with Matchers
+  with FutureAwaits
+  with DefaultAwaitTimeout
+  with IdiomaticMockito
+  with ArgumentMatchersSugar
+  with BeforeAndAfterAll
+  with DefaultPlayMongoRepositorySupport[PendingBalanceRequest] {
+
+  implicit val system       = ActorSystem(suiteName)
+  implicit val materializer = Materializer(system)
+  implicit val ec           = materializer.executionContext
+
+  val clock     = Clock.tickSeconds(ZoneOffset.UTC)
+  val lifecycle = new DefaultApplicationLifecycle
+  val random    = new SecureRandom
+
+  override def afterAll() = {
+    await(lifecycle.stop())
+    TestKit.shutdownActorSystem(system)
+    super.afterAll()
+  }
+
+  override lazy val repository = new BalanceRequestRepositoryImpl(
+    mongoComponent,
+    mkAppConfig(Configuration("mongodb.balance-requests.ttl" -> "5 minutes")),
+    clock,
+    random
+  )
+
+  def mkAppConfig(config: Configuration) = {
+    val servicesConfig = new ServicesConfig(config)
+    new AppConfig(config, servicesConfig)
+  }
+
+  val taxIdentifier      = TaxIdentifier("GB12345678900")
+  val guaranteeReference = GuaranteeReference("05DE3300BE0001067A001017")
+  val accessCode         = AccessCode("ABC1")
+
+  val balanceRequest = BalanceRequest(
+    taxIdentifier,
+    guaranteeReference,
+    accessCode
+  )
+
+  val balanceRequestSuccess = BalanceRequestSuccess(
+    BigDecimal("1212211848.45"),
+    CurrencyCode("GBP")
+  )
+
+  "BalanceRequestUpdateWorker" should "update the balance request cache when a request is updated" in {
+    val putBalanceDeferred = Deferred.unsafe[IO, Unit]
+    val completeDeferred   = putBalanceDeferred.complete(()).void
+
+    val cacheService = mock[BalanceRequestCacheService]
+
+    cacheService.putBalance(any[BalanceId], any[BalanceRequestResponse]) returns completeDeferred
+
+    new BalanceRequestUpdateWorker(
+      cacheService,
+      repository,
+      lifecycle
+    )
+
+    val balanceId = await {
+      repository.insertBalanceRequest(balanceRequest, clock.instant()).unsafeToFuture()
+    }
+
+    val response = await {
+      repository.updateBalanceRequest(balanceId.messageIdentifier, clock.instant(), balanceRequestSuccess)
+        .unsafeToFuture()
+    }
+
+    response shouldBe a[Some[_]]
+
+    await(putBalanceDeferred.get.unsafeToFuture())
+
+    val idCaptor       = ArgCaptor[BalanceId]
+    val responseCaptor = ArgCaptor[BalanceRequestResponse]
+
+    cacheService.putBalance(idCaptor.capture, responseCaptor.capture) wasCalled once
+
+    idCaptor.value shouldBe balanceId
+    responseCaptor.value shouldBe balanceRequestSuccess
+  }
```

Comment:
This test is rather flakey - it's usually successful when run with testOnly but fails much more frequently when a full test run is happening.
It's difficult to understand exactly what is happening because this fix relies on internal mechanisms of MongoDB (i.e. our code receiving Mongo change stream updates for the collection).
Any ideas about how to better test this would be much appreciated - perhaps I should use a fake mongo repository to supply the change stream?